File: 0124-binary-tree-maximum-path-sum/0124-binary-tree-maximum-path-sum.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {

    int max = Integer.MIN_VALUE;
    public int maxPathSum(TreeNode root) {
        if (root == null) return 0;
        helper(root, 0);
        return max;
    }

    private int helper(TreeNode root, int _max) {
        if (root == null) return 0;

        int left = helper(root.left, _max);
        int right = helper(root.right, _max);

        _max = root.val + Math.max(left, right);

        int localmax = Math.max(Math.max(Math.max(left, right), left + right + root.val), max);
        if (localmax != 0) max = localmax;
        else max = Math.max(max, _max);

        return (_max > 0) ? _max : 0;
    }
}

File: 0124-binary-tree-maximum-path-sum/0124-binary-tree-maximum-path-sum.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    int maxPathSum(TreeNode* root) {
        int res = INT_MIN;
        dfs(root, res);
        return res;
    }

    int dfs(TreeNode* root, int& res) {
        if (root == nullptr) return 0;

        int left = dfs(root->left, res);
        int right = dfs(root->right, res);

        int curr = root->val;
        if (left > 0) curr += left;
        if (right > 0) curr += right;

        res = std::max(res, curr);

        return std::max(0, std::max(left, right)) + root->val;

    }
};